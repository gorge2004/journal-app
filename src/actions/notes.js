import { db } from "../firabase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";
import { type } from "@testing-library/user-event/dist/type";
//react-journal
export const startNewNote = () => {
  return async (dispatch, getState) => {
    //get current state on redux
    const uid = getState().auth.uid;
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };
    try {
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      dispatch(activetNote(doc.id, newNote));
      dispatch(addNewNote(doc.id, newNote));
    } catch (error) {}
  };
};

export const activetNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id: id,
      ...note,
    },
  };
};
export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => {
  return {
    type: types.notesLoad,
    payload: notes,
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    for (const key in note) {
      if (note[key] === undefined) {
        delete note[key];
      }
    }
    try {
      const uid = getState().auth.uid;
      const noteToFirestore = { ...note };
      delete noteToFirestore.id;
      await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

      dispatch(refreshNote(note.id, noteToFirestore));

      Swal.fire("Saved", note.title, "success");
    } catch (error) {
    }
  };
};
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: { id, note: { id, ...note } },
});

export const startuploadingNote = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: "Uploading...",
      text: "Please Wait...",
      allowOutsideClick: false,
      showConfirmButton: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });
    const urlImage = await fileUpload(file);
    activeNote.url = urlImage;
    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = await getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
