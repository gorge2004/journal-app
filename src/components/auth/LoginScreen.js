import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "../../hooks/useForm";
import {
  startGoogleLogin,
  startLoginByEmailPassword,
} from "../../actions/auth";
import UIErrorForm from "../../ui/UIErrorForm";
import React from "react";
const LoginScreen = React.memo(() => {
  const { msgError, loading } = useSelector((state) => state.ui);
  const dispacth = useDispatch();
  const [{ email, password }, handleInputchange, reset] = useForm({
    email: "jorge@gmail.com",
    password: "123456",
  });

  const handleLogin = (evt) => {
    evt.preventDefault();
    dispacth(startLoginByEmailPassword(email, password));
  };
  const handleLoginByGoogle = (evt) => {
    dispacth(startGoogleLogin());
  };
  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form onSubmit={handleLogin} className="animate__animated animate__fadeIn animate__faster">
        { msgError &&  <UIErrorForm msg={msgError} />  }
        <input
          className="auth__input"
          autoComplete="off"
          type="text"
          placeholder="Email"
          name="email"
          value={email}
          onChange={handleInputchange}
        />
        <input
          className="auth__input"
          autoComplete="off"
          type="password"
          placeholder="password"
          name="password"
          value={password}
          onChange={handleInputchange}
        />

        <button className=" btn btn-primary btn-block" type="submit" disabled={loading}>
          Login
        </button>
        <div className="auth__social-networks">
          <p>Login with social network</p>
          <div className="google-btn" onClick={handleLoginByGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link className="link" to="/auth/register">
          Create a new account
        </Link>
      </form>
    </>
  );
});

export default LoginScreen;
