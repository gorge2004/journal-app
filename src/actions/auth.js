import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from "../firabase/firebase-config";
import { types } from "../types/types";
import { noteLogout } from './notes';
import { finishLoading, removeError, setError, startLoading } from "./ui";

//el middleware va recibir una accion , aqui se regresa una  funcion
export const startLoginByEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(removeError());
    dispatch(startLoading());
   return  firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(setError(err.message));
        dispatch(finishLoading());
        Swal.fire('ERROR',err.message, 'error');
      });
  };
};
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then((userCredential) => {
        const { uid, displayName } = userCredential.user;
        dispatch(login(uid, displayName));
      });
  };
};

export const startRegisterWithEmailAndPassword = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        console.log(user.uid, user.displayName);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        Swal.fire('ERROR',error.message, 'error');

      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout =  (email, password) => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    dispatch(logout());
    dispatch(noteLogout());
  };
};

export const logout = () => ({ type: types.logout});
