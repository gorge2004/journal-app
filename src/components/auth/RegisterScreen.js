import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { removeError, setError } from "../../actions/ui";
import { startRegisterWithEmailAndPassword } from "../../actions/auth";
const RegisterScreen = () => {
  const [{ name, email, password, password2 }, handleInputchange, reset] =
    useForm({
      name: "Jorge",
      email: "jorge@gmail.com",
      password: "123456",
      password2: "123456",
    });
  // obetengo el state del redux tal como se tiene
  const { msgError } = useSelector((state) => state.ui);
  const dispatch = useDispatch();
  const handleSubmitRegister = (evt) => {
    evt.preventDefault();
    if (isFormValue()) {
      console.log(name, email, password, password2);
      dispatch(startRegisterWithEmailAndPassword(email, password, name));
      
    }
  };
  const isFormValue = () => {
    if (name.trim().length === 0) {
      dispatch(setError("name is required."));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("email is not validate."));
      return false;
    } else if (password !== password2 || password2.length <= 5) {
      dispatch(
        setError("password should be at least characters and match each other.")
      );
      return false;
    }
    dispatch(removeError());

    return true;
  };
  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form onSubmit={handleSubmitRegister} className="animate__animated animate__fadeIn animate__faster">
        {msgError && <div className="auth__alert-error">{msgError}</div>}

        <input
          className="auth__input"
          autoComplete="off"
          type="text"
          placeholder="Name"
          name="name"
          value={name}
          onChange={handleInputchange}
        />
        <input
          className="auth__input"
          autoComplete="off"
          type="text"
          placeholder="Email"
          name="email"
          value={email}
          onChange={handleInputchange}
        />
        <input
          className="auth__input"
          autoComplete="off"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputchange}
        />
        <input
          className="auth__input"
          autoComplete="off"
          type="password"
          placeholder="Confirm password"
          name="password2"
          value={password2}
          onChange={handleInputchange}
        />
        <button className=" btn btn-primary btn-block mb-5" type="submit">
          Login
        </button>

        <Link className="link mt-5" to="/auth/login">
          already registered ?
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
