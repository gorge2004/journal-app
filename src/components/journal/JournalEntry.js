import React from "react";
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activetNote } from "../../actions/notes";

function JournalEntry({ id, date, url, title, body }) {
  const dispatch = useDispatch();
  const noteDate = moment(date);
  const handleClickSelect = () => {
    dispatch(activetNote(id,{  date, url, title, body }));
  }
  return (
    <div className="journal__entry animate__animated animate__fadeIn animate__faster pointer" onClick={handleClickSelect}>
      {url && (
        <div
          className="journal__entry-picture"
          style={{
            backgroundSize: "cover",
            backgroundImage: `url(${url})`,
          }}
        ></div>
      )}

      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>
        <p className="journal__entry-content">{body}</p>
      </div>
      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  );
}

export default JournalEntry;
