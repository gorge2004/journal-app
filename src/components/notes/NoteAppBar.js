import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startuploadingNote } from "../../actions/notes";

const NoteAppBar = () => {
  const fileRef = useRef();
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const handleSaveClick = () => {
    dispatch(startSaveNote(note));
  };
  const handlePictureClick = (evt) => {
    fileRef.current.click();
  };
  const handleFileChange = (evt) => {
    const file = evt.target.files[0];
    if (file) {
      dispatch(startuploadingNote(file));
    }
  };
  return (
    <div className="notes__appbar">
      <span>15 de febrero 2022</span>
      <input
        type="file"
        name="file"
        style={{
          display: "none",
        }}
        ref={fileRef}
        onChange={handleFileChange}
      />
      <div>
        <button className="btn" onClick={handlePictureClick}>
          Picture
        </button>
        <button className="btn" onClick={handleSaveClick}>
          Save
        </button>
      </div>
    </div>
  );
};

export default NoteAppBar;
