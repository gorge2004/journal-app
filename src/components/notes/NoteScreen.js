import { useEffect, useRef } from "react";
import NoteAppBar from "./NoteAppBar";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "../../hooks/useForm";
import { activetNote, startDeleting } from "../../actions/notes";


function NoteScreen() {

  const { active: note } = useSelector((state) => state.notes);
  const dispatch = useDispatch();
  const [formValues, handleInputChange, reset] = useForm(note);
  const { id, title, body } = formValues;
  const activeId = useRef(note.id);
  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
   
    dispatch(activetNote(id,{... formValues}));
  }, [formValues, dispatch]);
  
  const handleDeleteClick = () => {
    dispatch(startDeleting(id));
  }

  return (
    <div className="notes__main-content">
      <NoteAppBar />
      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="note__title-input"
          name="title"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="What happened today"
          className="note__textarea"
          onChange={handleInputChange}
          name="body"
          value={body}
        ></textarea>

        {note.url && (
          <div className="notes__image">
            <img
              src={note.url }
              alt="imagen"
            />
          </div>
        )}
      </div>
      <button className="btn btn-danger" onClick={handleDeleteClick}>
        Delete
      </button>
    </div>
  );
}

export default NoteScreen;
