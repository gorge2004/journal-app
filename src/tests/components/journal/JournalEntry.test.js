import JournalEntry from "../../../components/journal/JournalEntry";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { startLogout } from "../../../actions/auth";
import { activetNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const note = { id: "123s", title: "John", body: "This is john" };

const initialState = {
  auth: { uid: "123", name: "John" },
  ui: {},
  notes: { notes: [note], active: note },
  url: "http://localhost",
};
let store = mockStore(initialState);
store.dispatch = jest.fn(() => {});
const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...note} />
  </Provider>
);

describe("testing_JournalEntry", () => {
  test("should first", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("should_call_activeNote_Action", () => {
    wrapper.find(".journal__entry").simulate("click", {});
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(activetNote(note.id, {...note}));
  });
});
