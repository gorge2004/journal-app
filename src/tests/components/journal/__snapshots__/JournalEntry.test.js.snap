// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing_JournalEntry should first 1`] = `
<div
  className="journal__entry animate__animated animate__fadeIn animate__faster pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      John
    </p>
    <p
      className="journal__entry-content"
    >
      This is john
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Tuesday
    </span>
    <h4>
      1st
    </h4>
  </div>
</div>
`;
